include/bout/petsc_interface.hxx

```diff
@@ -0,0 +1,576 @@
+  }
+  friend void swap<F>(PetscMatrix<F>& first, PetscMatrix<F>& second);
+
+  PetscMatrixElement& operator()(ind_type& index1, ind_type& index2) {
+    int global1 = indexConverter->getGlobal(index1),
+        global2 = indexConverter->getGlobal(index2);
+#if CHECKLEVEL >= 1
+    if (!initialised) {
+      throw BoutException("Can not return element of uninitialised matrix");
+    } else if (global1 == -1 || global2 == -1) {
+      throw BoutException("Request to return invalid matrix element");
+    }
+#endif
+    std::vector<PetscInt> positions;
+    std::vector<PetscScalar> weights;
+    if (yoffset != 0) {
+      std::vector<ParallelTransform::positionsAndWeights> pw;
+      if (yoffset == -1) {
+	pw = pt->getWeightsForYDownApproximation(index2.x(), index2.y(), index2.z());
+      } else if (yoffset == 1) {
+	pw = pt->getWeightsForYUpApproximation(index2.x(), index2.y(), index2.z());
+      } else {
+	pw = pt->getWeightsForYApproximation(index2.x(), index2.y(), index2.z(), yoffset);
+      }
+      int ny = indexConverter->getMesh()->LocalNy, nz = indexConverter->getMesh()->LocalNz;
+      if (std::is_same<F, FieldPerp>::value) {
+	ny = 1;
+      } else if (std::is_same<F, Field2D>::value) {
+	nz = 1;
+      }
+      std::transform(pw.begin(), pw.end(), std::back_inserter(positions),
+		     [this, &ny, &nz](ParallelTransform::positionsAndWeights p) -> PetscInt
+		     {return this->indexConverter->getGlobal(ind_type(p.i*ny*nz + p.j*nz
+								      + p.k, ny, nz));});
+      std::transform(pw.begin(), pw.end(), std::back_inserter(weights),
+		     [](ParallelTransform::positionsAndWeights p) ->
+		     PetscScalar {return p.weight;});
+    }
+    return PetscMatrixElement::newElement(matrix.get(), global1, global2, positions, weights);
+  }
+
+  void assemble() {
+    MatAssemblyBegin(*matrix, MAT_FINAL_ASSEMBLY);
+    MatAssemblyEnd(*matrix, MAT_FINAL_ASSEMBLY);
+  }
+
+  void destroy() {
+    if (*matrix != nullptr && initialised) {
+      MatDestroy(matrix.get());
+      *matrix = nullptr;
+      initialised = false;
+    }
+  }
+
+  PetscMatrix<F> yup(int index = 0) {
+    return ynext(index + 1);
+  }
+  PetscMatrix<F> ydown(int index = 0) {
+    return ynext(-index - 1);
+  }
+  PetscMatrix<F> ynext(int dir) {
+    if (std::is_same<F, FieldPerp>::value && yoffset + dir != 0) {
+      throw BoutException("Can not get ynext for FieldPerp");
+    } else {
+      PetscMatrix<F> result; // Can't use copy constructor because don't
+                             // want to duplicate the matrix
+      result.matrix = matrix;
+      result.indexConverter = indexConverter;
+      result.pt = pt;
+      result.yoffset = std::is_same<F, Field2D>::value ? 0 : yoffset + dir;
+      result.initialised = initialised;
+      return result;
+    }
+  }
+
+  /// Provides a reference to the raw PETSc Mat object.
+  Mat* getMatrixPointer() {
+    return matrix.get();
+  }
+
+private:
+  shared_ptr<Mat> matrix;
+  IndexerPtr indexConverter;
```

Comment:
You use `shared_ptr` here, for the `matrix` and `indexConverter`, but just plain `Vec` in the `PetscVector`. Is there a reason, or are you just experimenting with the design?

Reply:
I used shared_ptr because I want to have multiple PetscMatrix objects corresponding to the same PETSc Mat object. This is necessary for the ynext() (and related) methods. I should probably use unique_ptr in PetscVector, but as it is easier to be memory-safe with a unique object than a shared one I didn't think to when I wrote it.